### PR TITLE
Wrong preprocessed query for nullable contraints

### PR DESCRIPTION
**Pull request by darthunix**
At first the problem was detected in Greenplum
```
create table ta(a int check(a = 1));
insert into ta values(null);

set optimizer=off;
select * from ta where a is null;
 a 
---
  
(1 row)

set optimizer=on;
select * from ta where a is null;
 a 
---
(0 rows)

set optimizer_print_query = on;
set client_min_messages='log';
select * from ta where a is null;
Algebrized query: 
+--CLogicalSelect
   |--CLogicalGet "ta" ("ta"), Columns: ["a" (0), "ctid" (1), "xmin" (2), "cmin" (3), "xmax" (4), "cmax" (5), "tableoid" (6), "gp_segment_id" (7)] Key sets: {[1,7]}
   +--CScalarNullTest
      +--CScalarIdent "a" (0)

Algebrized preprocessed query: 
+--CLogicalConstTableGet Columns: ["a" (0), "ctid" (1), "xmin" (2), "cmin" (3), "xmax" (4), "cmax" (5), "tableoid" (6), "gp_segment_id" (7)] Values: []
```
So ORCA preprocesses a logical plan to a single node  CLogicalConstTableGet, that means "there is no suitable data in this node, no need for retrieving data and constructing a physical plan". But in fact it is a wrong behavior for `ta` table - it can contain nulls as far as check `a = 1` doesn't protect from nulls. `select null = 1` returns null, but not a boolean value.

It was easy to guess that the problem was connected with constraint null checks and comparing with constants. Debugging showed out that CLogicalConstTableGet node is formed in [CExpressionPreprocessor.cpp](https://github.com/greenplum-db/gporca/blob/v3.65.2/libgpopt/src/operators/CExpressionPreprocessor.cpp#L1644) file and in a current example max cardinality is zero.

So here is a pipeline that caused the problem: [CLogicalSelect.cpp](https://github.com/greenplum-db/gporca/blob/v3.65.2/libgpopt/src/operators/CLogicalSelect.cpp#L141) -> [CConstraintInterval.cpp](https://github.com/greenplum-db/gporca/blob/v3.65.2/libgpopt/src/base/CConstraintInterval.cpp#L83)

As you can see the problem is in `m_fIncludesNull` parameter that describes whether nulls are allowed in a constraint or not. At the moment it is always false. But in fact it should be false only if a relative column not null - otherwise it should be true.

Under hood `m_fIncludesNull` is set in a function  [PciIntervalFromColConstCmp](https://github.com/greenplum-db/gporca/blob/v3.65.2/libgpopt/src/base/CConstraintInterval.cpp#L391) that transforms constraints comparing a column with a  constant to an interval one. So `a = 1` constraint always transforms to `a in [1, 1]` and `m_fIncludesNull` is always set false at the moment for a new interval constraint.

As a solution I propose to set `m_fIncludesNull` depending on whether a relative column is nullable or not. I am not sure that it solves the root of a problem but it seems to be a small change in a code and may be a good starting point.

Also this PR contains two tests: check constant constraint on a nullable and non-nullable column.
```
create table ta(a int check(a = 1));
insert into ta values(null);

create table tb(b int not null check(b = 1));
insert into tb values(1);
```
`ConstraintIntervatIncludesNull.mdp` was generated with a query that returned one null row with a TableScan: `select * from ta where a is null;` 

`ConstraintIntervalNotIncludesNull.mdp` was generated with a query that returned an empty output using OneTimeFilter (don't try to scan a table, there is nothing suitable there): `select * from tb where b is null;`

### CODE REVIEW

**Comment by darthunix** [comment]
To my great disappointment some of the tests have failed at the moment and I don't have access to view your internal Concourse output... If I run `ctest` on my laptop even on a master lots of tests fail... so it is very difficult to understand which of them have failed because of this commit

**Comment by bhuvnesh2703** [comment]
@darthunix thank you very much for reporting the fix, and even proposing a solution with detailed description of problem. Thats awesome 👏
Here is the list of tests which has failed for this PR:
```
The following tests FAILED:
	 43 - gporca_test_CSemiJoinTest (Failed)
	 44 - gporca_test_CAntiSemiJoinTest (Failed)
	 45 - gporca_test_CPredicateTest (Failed)
	 46 - gporca_test_CLikeIDFTest (Failed)
	 47 - gporca_test_CLeftOuterIndexApplyTest (Failed)
	 49 - gporca_test_CIndexScanTest (Failed)
	 50 - gporca_test_CIndexApplyTest (Failed)
	 51 - gporca_test_CDynamicIndexScanTest (Failed)
	 53 - gporca_test_CStatsTest (Failed)
	 54 - gporca_test_CICGMiscTest (Failed)
	 56 - gporca_test_CProjectTest (Failed)
	 57 - gporca_test_CPartTbl1Test (Failed)
	 58 - gporca_test_CPartTbl2Test (Failed)
	 59 - gporca_test_CPartTbl3Test (Failed)
	 60 - gporca_test_CPartTbl4Test (Failed)
	 61 - gporca_test_CPartTbl5Test (Failed)
	 63 - gporca_test_CSetop1Test (Failed)
	 64 - gporca_test_CSetop2Test (Failed)
	 65 - gporca_test_CSetop3Test (Failed)
	 66 - gporca_test_CSetop4Test (Failed)
	 69 - gporca_test_COuterJoin2Test (Failed)
	 73 - gporca_test_CMotionHazardTest (Failed)
	 80 - gporca_test_CReplicatedTableTest (Failed)
	 81 - gporca_test_CDqaTest (Failed)
	 94 - gporca_test_CWindowTest (Failed)
	 95 - gporca_test_CICGTest (SEGFAULT)
	 96 - gporca_test_CMultilevelPartitionTest (Failed)
	 97 - gporca_test_CDMLTest (Failed)
	 98 - gporca_test_CDirectDispatchTest (Failed)
	101 - gporca_test_CAggTest (Failed)
	102 - gporca_test_CSubqueryTest (Failed)
	106 - gporca_test_CBitmapTest (Failed)
	107 - gporca_test_CCTETest (Failed)
	109 - gporca_test_CPhysicalParallelUnionAllTest (Failed)
	110 - gporca_test_CMinidumpWithConstExprEvaluatorTest (SEGFAULT)
	124 - gporca_test_CTranslatorExprToDXLTest (Failed)
	132 - gporca_test_CCorrelatedExecutionTest (Failed)
	135 - gporca_test_CCastTest (Failed)
	136 - gporca_test_CConstTblGetTest (Failed)
	138 - gporca_test_CSubqueryHandlerTest (Failed)
	139 - gporca_test_CBindingTest (Failed)
	140 - gporca_test_CEngineTest (Failed)
	142 - gporca_test_CExpressionTest (Failed)
	151 - gporca_test_CPartConstraintTest (Failed)
Errors while running CTest
```
On our machine, no ctests fails locally (As dev practice we run the ctests locally everytime even before we create a PR, so we ensure that they pass locally).. So, there should be something wrong with the local environment you have. I have attached the file output from the pipeline.
[test_results.log](https://github.com/greenplum-db/gporca/files/3535778/test_results.log)

I have not reviewed the PR, but will do. Also a quick info, for many of the plan changes you can run the below script which will update the mdp. There will be some cases where it might not update, but should help.
```
ctest -j8 --rerun-failed --output-on-failure | tee /tmp/failures.out
../scripts/fix_mdps.py --logFile /tmp/failures.out
```

Here is the snippet from my local env which shows all ctests pass..
```
100% tests passed, 0 tests failed out of 161
```

**Comment by darthunix** [comment]
I'll make some investigations on the failed tests - sorry for pushing such a problem code

**Comment by darthunix** [comment]
After inspection of the broken tests I also agree that this PR is incorrect. As a result it breaks partition table tests as far as they rely on check constraints as well.
Thank you for your advice, tomorrow I'll inspect the part of the code you have mentioned

**Comment by darthunix** [comment]
I have changed the logic of setting max cardinality zero. Now it passes all local tests and solves the issue, but from my point of view the solution is ugly. May be you can advice me how to refactor this code it a better way?

**Comment by darthunix** [comment]
Hmm... Though the local tests have passed, concourse one failed. Can you show me the output of CI?

**Comment by darthunix** [comment]
I have rewritten the implementation of max cardinality zero check and added comments to minidump tests. Now PR passes all checks

**Comment by zellerh** [comment]
@darthunix, thanks for finding this bug!!

I liked your first fix (as far as I remember it was in `CConstraintInterval::PciIntervalFromColConstCmp()`?), but as @hardikar already said, thought it would have to be conditional, done only when we are evaluating a constraint expression and not when we process a selection predicate.

Unfortunately I didn't have the time to try this, but what about adding an extra bool parameter to `CConstraintInterval::PciIntervalFromColConstCmp()` and its callers, to tell us whether we are in a constraint or not. The parameter could be set in places like `CConstraintInterval::PciIntervalFromConstraint()` where we have that information.

So, something like the following:

```
CConstraintInterval *
CConstraintInterval::PciIntervalFromColConstCmp
	(
	...
	CScalarConst *popScConst,
	bool in_constraint
	)
{
	...
	if (NULL != pdrngprng)
	{
                 // if we are in a constraint, a NULL value of the column satisfies the constraint
		pcri = GPOS_NEW(mp) CConstraintInterval(mp, colref, pdrngprng, in_constraint /*fIncludesNull*/);
	}
	return pcri;

}
```
We may need similar logic for CConstraintInterval::PciComplement(). Example: A check constraint NOT (a = 1). The interval should again contain a NULL.

Here is a stack trace I'm seeing, with some suggestions for a fix:

```
...
  CLogical::PpcDeriveConstraintFromTable              // pass "true" to the new boolean parameter in PcnstrFromScalarExpr
   CConstraint::PcnstrFromScalarExpr                  // add a new "in_constraint" (or similar) parameter
      CConstraintInterval::PciIntervalFromColConstCmp // add a new "in_constraint" parameter
        CConstraintInterval::CConstraintInterval      // this receives "in_constraint" as the fIncludesNull parameter of this method
```

Again, sorry for not trying this fix myself first.

**Comment by darthunix** [comment]
I have changed function API through the stack and added bool flag `in_constraint` with default false value (as I have been adviced many times in this PR - I didn't want to change API at first, sorry...)
```
CLogical::PpcDeriveConstraintFromTable
CConstraint::PcnstrFromScalarExpr
CConstraintInterval::PciIntervalFromScalarExpr
CConstraintInterval::PciIntervalFromScalarCmp
CConstraintInterval::PciIntervalFromColConstCmp
CConstraintInterval::CConstraintInterval
```

At the moment we don't need any modifications in  CConstraintInterval::PciComplement() as far as check constraint with `NOT (a = 1)` doesn't reproduce an error after patch.

**Comment by hardikar** [comment]
@darthunix What you have now LGTM. There are some renames & comment move around that I wanted to do, so I just pushed another commit to the PR. The logic is the same, just different naming.

We can take it up from here to merge this into ORCA. We need to go through some internal process & also write up tests for GPDB. Thanks again for report the bug and working on fixing it!

**Comment by hardikar** [comment]
As a side note, while looking into this, I did find a different case that is still broken:
```
create table td (a int check (a IN (1, 2)));
insert into td values (null);
select * from td where a is null;
 a
---
(0 rows)
shardikar=# set optimizer =off;
SET
shardikar=# select * from td where a is null;
 a
---

(1 row)
```

I'm working on a fix for that.

**Comment by darthunix** [comment]
Sorry for keeping silence for several days - I am on my vacations at the moment. 
I found out that nullable checks are still broken in a case of `and`/`or` conditions.
For example, `a != 1` check constraint works after the patch, but `a != 1 and a != 3` doesn't. The same problem with `a between 1 and 3` (under hood it is `a >=1 and a <=3`). Also `or` condition doesn't work as well - `a != 1 or a != 3` returns zero rows with null filter.
```
postgres=# create table ta(a int check(a != 1 or a != 3));
postgres=# insert into ta values(null);
INSERT 0 1
postgres=# select * from ta where a is null;
 a 
---
(0 rows)
```
I'll try to send a fix

**Comment by darthunix** [comment]
I have fixed IN and boolean operators scenario (it seems we have covered all from CConstraintInterval::PciIntervalFromScalarExpr, right?) + added tests.

File: libgpopt/include/gpopt/base/CConstraint.h [hunk 1]
```diff
@@ -156,7 +156,13 @@ namespace gpopt
 			// create constraint from scalar array comparison expression originally generated for
 			// "scalar op ANY/ALL (array)" construct
 			static
-			CConstraint *PcnstrFromScalarArrayCmp(CMemoryPool *mp, CExpression *pexpr, CColRef *colref);
+			CConstraint *PcnstrFromScalarArrayCmp
+							(
+							CMemoryPool *mp,
+							CExpression *pexpr,
+							CColRef *colref,
+							BOOL infer_nullability = true
```
Comment (chrishajas) [inline libgpopt/include/gpopt/base/CConstraint.h]: Do we want `infer_nullability` to have a default value here and elsewhere in this PR? It seems like it could be error-prone if we forget to include it

**Comment by hardikar** [comment]
Thanks @darthunix for add commits for AND/OR and IN/NOT IN cases. There is actually another scenario which needs to be handled also, when `set optimizer_array_constraints=on`. I added a commit that fixes that. Looks like it needed code change to fix an existing issue also.

Anyway, it's looking pretty good now. I took a look into why we don't need to touch NOT. That's because of way we derive the constraint for NOT operators. We derive the constraint for the sub expression and then "complement" it. Thus, if NULL wasn't included in the inner expression, we'd add it. E.g For `NOT (a =5)`, `a=5` => `[5,5] NULL excluded`. Complement that to get `(-inf, 5], [5, inf), NULL included`. This is precarious, but I'm still trying to come with an example where it'll fail. IS NOT DISTINCT FROM doesn't suffer from this (I think).

File: libgpopt/include/gpopt/base/CConstraint.h (hunk 1 above)
Comment (hardikar) [reply libgpopt/include/gpopt/base/CConstraint.h]: I personally don't like default positional arguments like this either. But 90% of the cases, we want it to be `true`, a.k.a some default behavior.

File: libgpopt/src/base/CConstraintInterval.cpp [hunk 2]
```diff
@@ -293,7 +295,7 @@ CConstraintInterval::PcnstrIntervalFromScalarArrayCmp
 		}
 	}
 
-	BOOL fContainsNull = apdatumsortedset->FIncludesNull();
+	BOOL fContainsNull = !infer_nullability;
```
Comment (ashuka24) [inline libgpopt/src/base/CConstraintInterval.cpp]: here we use `fContainsNull` and everywhere else we use some variation of `include_null`,`fIncludesNull`,etc , can we update this here for consistency?

File: libgpopt/src/base/CConstraintInterval.cpp [hunk 3]
```diff
@@ -544,10 +554,10 @@ CConstraintInterval::PciIntervalFromScalarBoolOp
...
 		case CScalarBoolOp::EboolopAnd:
-			return PciIntervalFromScalarBoolAnd(mp, pexpr, colref);
+			return PciIntervalFromScalarBoolAnd(mp, pexpr, colref, infer_nullability);
 
 		case CScalarBoolOp::EboolopOr:
-			return PciIntervalFromScalarBoolOr(mp, pexpr, colref);
+			return PciIntervalFromScalarBoolOr(mp, pexpr, colref, infer_nullability);
 
 		case CScalarBoolOp::EboolopNot:
 		{
```
[the remark this answers is not in the document]
Comment (darthunix) [reply libgpopt/src/base/CConstraintInterval.cpp]: @zellerh at the moment NULL logic gives correct result (though I am not sure it works right).
1. Lets imagine we have a check constraint `not (a = 5)`
2. In `PciIntervalFromScalarBoolOp` function we choose `EboolopNot` case option and create from `a = 5` an interval `{ [5] }` without NULL (as far as `infer_nulls_as = false` by default)
3. After that we call `PciComplement` for an interval `{ [5] }` to invert it
   - At first step we create a new inbounded interval with NULL { (-inf, +inf), [NULL] }
   - For unbounded interval with NULL we call `PciDifference` method, subtracting `{ [5] }`, i.e. `{ (-inf, +inf), [NULL] } - { [5] } = { (-inf, 5), (5, +inf), [NULL] }`

The problem is that for a check constraint `a = 5` `PciIntervalFromScalarExpr` would produce `{ [5], [NULL] }`. But for a check constraint `not (a = 5)` we get `{ [5] }` interval without NULL and invert it... A logic when we get in one case `{ [5] }` or `{ [5], [NULL] }` from one expression `a = 5` in different constraints doesn't look idempotent.

**Comment by hardikar** [comment]
Consider the following two tables:
```
create table te ( a int check (not (a != 5)));
create table td ( a int check (not (a is distinct from 5)));
```

For both, the constraints before applying the NOT should be `(-inf, 5), (4, +inf), NULL`. But after applying the NOT, it should be `[5,5], NULL` for te and `[5,5]` for td. That's the underlying problem since whether or not NULL is included makes a difference for queries that contain a `IS NULL` predicate. 

I was able to find a failing case because we don't pass `infer_null_as` to NOT:
```
create table tx ( a int check (not( not (a = 5) and not (a = 6))));
insert into tx values (null);
select * from tx where a is null
```

So, we should probably fix this.

As an alternative fix, I was thinking, what if we *always* include NULLs in constraint interval derived from check constraints. We will sometimes miss out some optimization. Specifically, we would have to scan the table for IS NULL predicate when the check constraint had a IS NOT NULL and/or IS DISTINCT FROM in it. But we will never get wrong results. Note that this would only apply for IS NULL predicates. If the table has a column `a int check (a < 5)` and the query predicate has `a > 5`, ORCA can still convert that into Result (false).

@zellerh What do you think about this solution to get around the constraint framework as it stands now?

The alternative I see is to introduce a new comparator type: NIDF, in line with `COperator::EopScalarIsDistinctFrom` & `ECmpType::EcmptIDF`, and change ORCA's handling thereof.

Thoughts?

**Comment by hardikar** [comment]
Or course, we could still merge it as is now, and come back to handle NOT in the future. But it does make me uneasy that it could produce wrong results. I'd rather have slow plans than wrong ones.

**Comment by darthunix** [comment]
I have modified `PciComplement` method - now it can preserve NULLs on demand (with a bool flag) when we create an inverted interval with NOT. I am not sure that it is a good solution as far as we have a table scan for IS DISTINCT FROM and IS NOT NULL. May be a better one would be an additional logic for a new comparator type: NIDF...

File: data/dxl/minidump/ConstraintIntervalNotIncludesNull.mdp [hunk 4]
```diff
@@ -0,0 +1,248 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
+<dxl:Comment><![CDATA[
+create table tb(b int not null check(b = 1));
+insert into ta values(1);
```
[the remark this answers is not in the document]
Comment (darthunix) [reply data/dxl/minidump/ConstraintIntervalNotIncludesNull.mdp]: Fixed

**Comment by hardikar** [comment]
@darthunix Thanks so much for pushing fixes for this still!

As an aside, you can find out what ORCA is falling back by `set client_min_messages = 'log';`. For me (with assets enabled), it fails on `../libgpopt/src/operators/CLogical.cpp:796: Failed assertion: CUtils::FPredicate(pexprCheckConstraint)`.

**Comment by hardikar** [comment]
I also checked to see that your PR works fine for: `create table t7 ( a int check not (a is not distinct from 5 or a is not distinct from 6));` case also!

```
create table t1 (a int check (a is not distinct from 6);
explain select * from t1 where a is not null;   -- uses scan instead of result(false)
```
I'm OK with that, as far as this PR is concerned, because it still produces right results.

**Comment by hardikar** [comment]
Just pushed a fix that was causing that fallback.
```
shardikar=# explain select * from t5 where a is null;
                      QUERY PLAN
------------------------------------------------------
 Result  (cost=0.00..0.00 rows=0 width=4)
   ->  Result  (cost=0.00..0.00 rows=0 width=4)
         One-Time Filter: false
 Optimizer: Pivotal Optimizer (GPORCA) version 3.67.0
(4 rows)
```

File: data/dxl/minidump/ConstraintIntervalWithNotIncludesNull.mdp [hunk 5]
```diff
@@ -0,0 +1,259 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
+<dxl:Comment><![CDATA[
+create table tx (a int check (not(not(a = 5) and not(a = 6))));
+insert into tx values (null);
+explain select a from td where a is null;
```
Comment (chrishajas) [inline data/dxl/minidump/ConstraintIntervalWithNotIncludesNull.mdp]: td->tx

**Comment by zellerh** [comment]
Here are some thoughts on what those methods mean and how to express the code in a mathematical way:

Given a boolean expression _x_ on a colref _c_, generate the ranges (intervals) _r(x)_ for colref _c_ for which the predicate evaluates to true, or _r’(x)_ for which the expression evaluates to true or null. In the code, _r_ is the result with `infer_nulls_as = false`, _r'_ is the result with `infer_nulls_as = true`.

Allowed operations on ranges: intersect, union, complement

Rules:


Range of a Boolean expression | Equivalent | Comment
-- | -- | --
r(x and y) | r(x) intersect r(y) | Both x and y must be true for a value of c to qualify
r(x or y) | r(x) union r(y) | One of x or y must be true for a value of c to qualify
r(not x) | complement(r’(x)) | x must be false
r’(x and y) | r’(x) intersect r’(y) | Both x and y must not be false
r’(x or y) | complement(complement(r’(x)) intersect complement(r’(y))) | x and y must not both be false
r’(not x) | complement (r(x)) | x must not be true

Truth table:

x | y | x and y | x or y | not x | where x and y | where x or y | where not x | check x and y | check x or y | check not x
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
false | false | false | false | true |   |   | X |   |   | X
false | true | false | true | true |   | X | X |   | X | X
false | null | false | null | true |   |   | X |   | X | X
true | false | false | true | false |   | X |   |   | X |  
true | true | true | true | false | X | X |   | X | X |  
true | null | null | true | false |   | X |   | X | X |  
null | false | false | null | null |   |   |   |   | X | X
null | true | null | true | null |   | X |   | X | X | X
null | null | null | null | null |   |   |   | X | X | X

Would this give us a good guidance for implementation?

Take the example above,

```
create table te ( a int check (not (a != 5)));
create table td ( a int check (not (a is distinct from 5)));
```
Applying the rules above:

```
r'(not x) = complement(r(x))
```

For table te:

```
r(a != 5) = { (-inf, 5), (5, +inf) }
complement(r(a != 5)) = { [5], null }
```

For table td:

```
r(a is distinct from 5) = { (-inf, 5), (5, +inf), null }
complement(r(a is distinct from 5)) = { [5] }
```

**Comment by zellerh** [comment]
Thanks, @hardikar, for letting me know that I can push to this branch, so I went ahead and added the suggested commit to this PR.

**Comment by darthunix** [comment]
I have added multicolumn check constraint support (we returned wrong results for `check( a + b > 1 and a = b)` for example).

File: data/dxl/minidump/ConstraintIntervalWithNotIncludesNull.mdp (hunk 5 above)
Comment (darthunix) [reply data/dxl/minidump/ConstraintIntervalWithNotIncludesNull.mdp]: Done

File: libgpopt/src/base/CConstraint.cpp [hunk 6]
```diff
@@ -207,12 +209,12 @@ CConstraint::PcnstrFromScalarExpr
...
+		pcnstr = CConstraintInterval::PciIntervalFromScalarExpr(mp, pexpr, colref, infer_nullability);
 		if (NULL == pcnstr && CUtils::FScalarArrayCmp(pexpr))
 		{
 			// if the interval creation failed, try creating a disjunction or conjunction
 			// of several interval constraints in the array case
-			pcnstr = PcnstrFromScalarArrayCmp(mp, pexpr, colref);
+			pcnstr = PcnstrFromScalarArrayCmp(mp, pexpr, colref, infer_nullability);
 		}
 
 		if (NULL != pcnstr)
```
[the remark this answers is not in the document]
Comment (darthunix) [reply libgpopt/src/base/CConstraint.cpp]: Done

**Comment by darthunix** [comment]
Have we fixed all the possible problems about nulls in check constraints? As for me I don't see any more... 
What else is required to finish with this PR? I remember about:

- https://github.com/greenplum-db/gpdb/pull/8606 (regression tests for ORCA)
  As far as I understand we have to extend regression tests with multicolumn scenarios and add one-time filter tests as well for `not null` columns (the same logic as in a current PR tests).
- https://github.com/greenplum-db/gpdb/issues/8582 (Postgres planner issue same as in a current PR)
  Is it ok to merge this PR in ORCA when Postgres planner returns different (and wrong) results? Or we have to fix Postgres planner as well before merge?

And one more question - what is the policy in ORCA for commits? Should we squash them all in one big commit per PR or is it normal to have lots of small ones?

**Comment by zellerh** [comment]
Hi Denis,

Agree with you, the PRs should hopefully be ready to be merged. I would not
wait for the planner fix. Having ORCA give the correct result provides a
workaround for people who run into the planner bug.

Yes, when we merge the PR we will need to squash all the commits into one.
Before we can do that, we will need to have approvals from more people.

What kind of multi-column tests where you thinking about? You added one MDP
that has multiple columns (
https://github.com/greenplum-db/gporca/pull/527/files#diff-817ca888858dce7563940ca871ce30ce),
would that be enough?

Thanks,

Hans

**Comment by darthunix** [comment]
Hi, Hans!
Yes, I’ve meant the test you mentioned and its variant with one-time filter - https://github.com/darthunix/gporca/blob/99d2557025b7441ff82d6c06cbf81be699f99dd8/data/dxl/minidump/ConstraintIntervalNotIncludesNull.mdp

**Comment by darthunix** [comment]
Btw, can you squash the commits yourself in a current PR, or it should be done by me?

**Comment by zellerh** [comment]
@darthunix, I'll be squashing and merging this for you. Thank you very much again for your contribution!

**Comment by zellerh** [comment]
Sorry @darthunix, I had to force-push over your branch to resolve merge conflicts. If anyone wants to see the original commit history, see https://github.com/zellerh/gporca/commits/ADBDEV-288.
